Mage.Sets/src/mage/cards/j/JeskaiInfiltrator.java

```diff
@@ -81,38 +70,31 @@ public JeskaiInfiltratorEffect copy() {
-            cardsToManifest.add(source.getSourcePermanentIfItStillExists(game));
-            cardsToManifest.add(controller.getLibrary().getFromTop(game));
-            UUID exileId = UUID.randomUUID();
-            controller.moveCardsToExile(cardsToManifest, source, game, false, exileId, "");
-            ExileZone exileZone = game.getExile().getExileZone(exileId);
-            for (Card card : exileZone.getCards(game)) {
-                card.setFaceDown(true, game);
-            }
-            game.fireUpdatePlayersEvent(); // removes Jeskai Infiltrator from Battlefield, so Jeskai Infiltrator returns as a fresh permanent to the battlefield with new position
-
-            Ability newSource = source.copy();
-            newSource.setWorksFaceDown(true);
-            //the Set will mimic the Shuffling
-            exileZone.getCards(game).forEach(card -> {
-                ManaCosts manaCosts = null;
-                if (card.isCreature(game)) {
-                    manaCosts = card.getSpellAbility() != null ? card.getSpellAbility().getManaCosts() : null;
-                    if (manaCosts == null) {
-                        manaCosts = new ManaCostsImpl<>("{0}");
-                    }
-                }
-                Card battlefieldCard = BecomesFaceDownCreatureEffect.findDefaultCardSideForFaceDown(game, card);
-                MageObjectReference objectReference = new MageObjectReference(battlefieldCard.getId(), battlefieldCard.getZoneChangeCounter(game) + 1, game);
-                game.addEffect(new BecomesFaceDownCreatureEffect(manaCosts, objectReference, Duration.Custom, FaceDownType.MANIFESTED), newSource);
-            });
-            controller.moveCards(exileZone.getCards(game), Zone.BATTLEFIELD, source, game, false, true, false, null);
-            return true;
+        if (controller == null) {
+            return false;
+        }
+        UUID exileId = UUID.randomUUID();
+        Permanent sourcePermanent = source.getSourcePermanentIfItStillExists(game);
+        if (sourcePermanent != null) {
+            controller.moveCardsToExile(sourcePermanent, source, game, false, exileId, "");
+        }
+        Card topCard = controller.getLibrary().getFromTop(game);
+        if (topCard != null) {
+            controller.moveCardsToExile(topCard, source, game, false, exileId, "");
+        }
+        // need to get source permanent as card rather than permanent for next steps, hence this convoluted code
+        List<Card> cardsToManifest = new ArrayList<>(game.getExile().getExileZone(exileId).getCards(game));
+        if (cardsToManifest.isEmpty()) {
+            return false;
+        }
+        for (Card card : cardsToManifest) {
+            card.setFaceDown(true, game);
         }
-        return false;
+        Collections.shuffle(cardsToManifest);
+        game.informPlayers(controller.getLogName() + " shuffles the face-down pile");
+        game.getState().processAction(game);
```

Comment:
I'm not sure is it require processaction here, cause whole effect is a single action and nothing can trigger for a cards in exile before it moved next.

Reply:
I think it's important since permanent leaving battlefield then returning - similar to `ExileThenReturnTargetEffect` where it's definitely needed (possible bug: state triggered abilities must see the moment where the card is in exile and not on battlefield)